src/vtokenUpgradesHelper.ts

```diff
@@ -0,0 +1,157 @@
+    shortfall,
+    supplyRatePerBlock,
+  };
+};
+
+export const fetchVTokenStorageCore = async (vToken: ethers.Contract, user: string) => {
+  const name = await vToken.name();
+  const symbol = await vToken.symbol();
+  const decimals = await vToken.decimals();
+  const owner = await vToken.admin();
+  const comptroller = await vToken.comptroller();
+  const interestRateModel = await vToken.interestRateModel();
+  const reserveFactorMantissa = await vToken.reserveFactorMantissa();
+  const accrualBlockNumber = await vToken.accrualBlockNumber();
+  const borrowIndex = await vToken.borrowIndex();
+  const totalBorrows = await vToken.totalBorrows();
+  const totalSupply = await vToken.totalSupply();
+  const totalReserves = await vToken.totalReserves();
+  const underlying = await vToken.underlying();
+  const accountBalance = await vToken.callStatic.balanceOf(user);
+  const borrowBalance = await vToken.callStatic.borrowBalanceStored(user);
+  const borrowRatePerBlock = await vToken.borrowRatePerBlock();
+  const pendingOwner = await vToken.pendingAdmin();
+  const supplyRatePerBlock = await vToken.supplyRatePerBlock();
+
+  return {
+    name,
+    symbol,
+    decimals,
+    owner,
+    comptroller,
+    interestRateModel,
+    reserveFactorMantissa,
+    accrualBlockNumber,
+    borrowIndex,
+    totalBorrows,
+    totalSupply,
+    totalReserves,
+    underlying,
+    accountBalance,
+    borrowBalance,
+    borrowRatePerBlock,
+    pendingOwner,
+    supplyRatePerBlock,
+  };
+};
+
+export const performVTokenBasicActions = async (
+  marketAddress: string,
+  user: SignerWithAddress,
+  mintAmount: BigNumber,
+  borrowAmount: BigNumber,
+  repayAmount: BigNumber,
+  redeemAmount: BigNumber,
+  vToken: ethers.Contract,
+  underlying: ethers.Contract,
+  isUnderlyingMock: boolean,
+) => {
+  const underlyingDecimals = await underlying.decimals();
+  const symbol = await underlying.symbol();
+
+  if (symbol === "WBNB") {
+    mintAmount = parseUnits("1", 18);
+    borrowAmount = parseUnits("0.5", 18);
+    repayAmount = parseUnits("0.25", 18);
+    redeemAmount = parseUnits("0.5", 18);
+    await underlying.connect(user).deposit({ value: mintAmount });
+  }
+  if (underlyingDecimals == 6) {
+    mintAmount = parseUnits("200", 6);
+    borrowAmount = parseUnits("50", 6);
+    repayAmount = parseUnits("25", 6);
+    redeemAmount = parseUnits("50", 6);
+  }
+  if (process.env.FORK_TESTNET === "true" && isUnderlyingMock) {
+    try {
+      await underlying.connect(user).faucet(mintAmount.add(repayAmount));
+    } catch (error) {
+      await underlying.connect(user).allocateTo(user.address, mintAmount.add(repayAmount));
+    }
+  }
+
+  await underlying.connect(user).approve(marketAddress, mintAmount.add(repayAmount));
+
+  await vToken.connect(user).mint(mintAmount);
+  await vToken.connect(user).borrow(borrowAmount);
+  await vToken.connect(user).repayBorrow(repayAmount);
+  await vToken.connect(user).redeemUnderlying(redeemAmount);
```

Comment:
Shouldn't we check the responses are 0 (NO_ERROR)?

Reply:
I checked increase and decrease in balances of user [af09fb0](https://github.com/VenusProtocol/vips/pull/67/commits/af09fb00273c1ff846a5c74d45c6c17b0d6216b9)